### PR TITLE
Centralized term image url function & added filter

### PR DESCRIPTION
**Pull request by ttodua**
### All Submissions:

* [ x ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [ x ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ x ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes # .

### How to test the changes in this Pull Request:

1. For example, we can now add a hook to change the term images.
```
add_filter( 'woocommerce_get_term_src_url', 'example_func', 10, 5);
function example_func($image_url, $term_id, $size, $placeholder_image ) {
	...
	if ( condition ) 
		$image_url =  'https://i.vimeocdn.com/portrait/25122243_300x300';
	...
	return $image_url ;
}
```

### Other information:

* [x ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x ] Have you written new tests for your changes, as applicable?
* [x ] Have you successfully ran tests with your changes locally?

### Changelog entry

> Enter a short summary of all changes on this Pull Request. This will appear in the changelog if accepted.

1. I've Added a global function to get term thumbnail urls, so we can use this function, instead of repeated code for getting the image of term. This also has an essential hook for 3rd party customizations, which lacks current WC version for term images, because in some cases, developers (like me) use custom images for taxonomies (not uploaded inside standard WP uploader, like manual directories or CDN-ed files) to serve custom urls instead of native thumbnail urls.

2. Replaced all manually-repeated code in other places, with this function.

Now, many developers (including my firm) will have an useful way to change term images.

### CODE REVIEW

File: includes/wc-term-functions.php [hunk 1]
```diff
@@ -791,3 +791,34 @@ function wc_get_product_visibility_term_ids() {
...
+/**
+ * Get a thumbnail image for term.
+ *
+ * The returned variable is filtered by woocommerce_get_term_thumbnail_url filter to
+ * allow 3rd party customisation.
+ *
+ * @param array|int $term Term object of the target term, or term ID.
+ * @return string Final image url.
+ */
+function wc_get_term_src_url( $term, $image_id = false, $size = false, $placeholder_image = false ) {
```
Comment (mikejolley) [inline includes/wc-term-functions.php]: Can we name this `wc_get_term_image_url` so it's more similar to `wp_get_attachment_image_url`. Args should simply be $term_id and $size. The others are not needed.

You may need to add an `wc_get_term_image_id` also for code which needs ID rather than src.

File: includes/wc-term-functions.php [hunk 2]
```diff
@@ -791,3 +791,34 @@ function wc_get_product_visibility_term_ids() {
...
+	
+	$term_id = is_numeric($term) ? $term : $term->term_id;
+	$image_url = '';
+	
+	// Get term image id.
+	if ( ! $image_id )
```
Comment (mikejolley) [inline includes/wc-term-functions.php]: This does not follow coding standards.

File: includes/wc-term-functions.php [hunk 3]
```diff
@@ -791,3 +791,34 @@ function wc_get_product_visibility_term_ids() {
...
+		$image_id = (int) get_woocommerce_term_meta( $term_id, 'thumbnail_id' );
+	
+	// If image exists, get it
+	if ( $image_id ) {
+		$image_url = $size == 'thumb' ? wp_get_attachment_thumb_url( $image_id ) : wp_get_attachment_url( $image_id );
+	}
+	
+	// If placeholder image is needed
+	if ( empty( $image_url ) && $placeholder_image ) {
```
Comment (mikejolley) [inline includes/wc-term-functions.php]: Placeholder should not be optional. This function should return a src regardless.

File: includes/api/v2/class-wc-rest-product-categories-v2-controller.php [hunk 4]
```diff
@@ -62,11 +62,14 @@ public function prepare_item_for_response( $item, $request ) {
 				'date_created_gmt'  => wc_rest_prepare_date_response( $attachment->post_date_gmt ),
 				'date_modified'     => wc_rest_prepare_date_response( $attachment->post_modified ),
 				'date_modified_gmt' => wc_rest_prepare_date_response( $attachment->post_modified_gmt ),
-				'src'               => wp_get_attachment_url( $image_id ),
+				'src'               => wc_get_term_src_url( $item, $image_id, false, false ),
 				'title'             => get_the_title( $attachment ),
 				'alt'               => get_post_meta( $image_id, '_wp_attachment_image_alt', true ),
 			);
 		}
+		else{
```
Comment (mikejolley) [inline includes/api/v2/class-wc-rest-product-categories-v2-controller.php]: This does not follow coding standards.

File: includes/api/class-wc-rest-product-categories-controller.php [hunk 5]
```diff
@@ -62,11 +62,14 @@ public function prepare_item_for_response( $item, $request ) {
 				'date_created_gmt'  => wc_rest_prepare_date_response( $attachment->post_date_gmt ),
 				'date_modified'     => wc_rest_prepare_date_response( $attachment->post_modified ),
 				'date_modified_gmt' => wc_rest_prepare_date_response( $attachment->post_modified_gmt ),
-				'src'               => wp_get_attachment_url( $image_id ),
+				'src'               => wc_get_term_src_url( $item, $image_id, false, false ),
 				'name'              => get_the_title( $attachment ),
 				'alt'               => get_post_meta( $image_id, '_wp_attachment_image_alt', true ),
 			);
 		}
+		else{
```
Comment (mikejolley) [inline includes/api/class-wc-rest-product-categories-controller.php]: This does not follow coding standards - same mistake in multiple files.

**Review by mikejolley — CHANGES_REQUESTED** [review]

File: includes/api/class-wc-rest-product-categories-controller.php (hunk 5 above)
Comment (tazotodua) [reply includes/api/class-wc-rest-product-categories-controller.php]: Can you point me which area doesn't follow standards? commas, separated with spaces, etc... ?

File: includes/wc-term-functions.php (hunk 1 above)
Comment (tazotodua) [reply includes/wc-term-functions.php]: Mike, you are free to modify this as you think it will be better. 
Or I have to modify it?

File: includes/api/class-wc-rest-product-categories-controller.php (hunk 5 above)
Comment (mikejolley) [reply includes/api/class-wc-rest-product-categories-controller.php]: See https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#brace-style

File: includes/api/v2/class-wc-rest-product-categories-v2-controller.php (hunk 4 above)
Comment (tazotodua) [reply includes/api/v2/class-wc-rest-product-categories-v2-controller.php]: Mike, sorry for bothering, can you modify it as it needs to be? Because I am confused..

File: includes/wc-term-functions.php (hunk 1 above)
Comment (tazotodua) [reply includes/wc-term-functions.php]: What if we leave the other arguments? They are important, because if you carefully view  my other commited files, there are cases when it RELIES on those parameters.

File: includes/wc-term-functions.php (hunk 3 above)
Comment (tazotodua) [reply includes/wc-term-functions.php]: The problem is that the function in itself, are called from multiple places, and some of those areas DO require the placeholder image, and some of those areas (like REST API) doesn't require it (instead, they demand that if there is no image, then Placeholder image should not be returned.)

Probably you haven't look into other above commits, where those use-cases show the differences.

File: includes/api/v2/class-wc-rest-product-categories-v2-controller.php (hunk 4 above)
Comment (tazotodua) [reply includes/api/v2/class-wc-rest-product-categories-v2-controller.php]: You pointed me to the "braces" part in WP Coding Standards, but i don't see any braces related to these parts.. ( 
Probably it's not long when I started GitHub and I have problem to understand which parts you point me..

File: includes/api/class-wc-rest-product-categories-controller.php (hunk 5 above)
Comment (tazotodua) [reply includes/api/class-wc-rest-product-categories-controller.php]: I've updated my forked files you told me (Now braces uses the standards as it was said).

**Comment by tazotodua** [comment]
I think I've ammended with all your suggestions. However, with my unfamiliarity to GitHub I can't see what to do now.

**Review by ttodua — COMMENTED** [review]
I've amended the required changes.

**Comment by mikejolley** [comment]
Hmm you seem to have disregarded some of the feedback I provided. I don't think it makes sense to pass thumbnail IDs to this method, thats why I suggested making another function to get the thumbnail ID which `wc_get_term_image_url` can use given the $term it is given...

Current solution is a bit all over the place. `class-wc-admin-taxonomies.php` shouldn't be querying the thumbnail ID, `wc_get_term_image_url` should.

**Comment by tazotodua** [comment]
@mikejolley  I have not disregarded, as I've said, I am newcomer to GitHub, probably I didn't do something correctly. I just asked you, if you can modify it, then modify and merge it as you think to be better.

**Comment by mikejolley** [comment]
I'll put a new PR up soon. The API change to add image_src is not possible however; you cannot change the API schema like this without a new API version.

**Comment by mikejolley** [comment]
@tamarazuk After delving into this further, this isn't going to be an approuch we can accept for the following reasons:

1. Some code requires an attachment/attachment ID whearas some grabs the image URL. You code only deals with image URLs
2. API changes are not permitted without a major API update. The API itself requires an attachment to generate more information about the term image. Your function cannot account for that, and won't work with remote images.
3. The change to template functions does not implement this function, it uses a filter. This is inconsistent.

Remote images and attachments are not two things which mesh nicely. If you want to use remote images you can add new filters, like your suggested change in template-functions, to change the actual image with a remote one. This would need to be done everywhere the term image is output however, and not in the REST API.

tl;dr we cannot officially support remote images, but could accept new filters only to change HTML where images are rendered.

Thanks

cc @woocommerce/proton
